File: core/java/bestbefore-processor-java/src/main/java/com/github/gotson/bestbefore/processor/DeprecatedProcessor.java
package com.github.gotson.bestbefore.processor;

import com.google.auto.service.AutoService;

import javax.annotation.processing.AbstractProcessor;
import javax.annotation.processing.Processor;
import javax.annotation.processing.RoundEnvironment;
import javax.annotation.processing.SupportedAnnotationTypes;
import javax.annotation.processing.SupportedSourceVersion;
import javax.lang.model.SourceVersion;
import javax.lang.model.element.Element;
import javax.lang.model.element.TypeElement;
import javax.tools.Diagnostic;
import java.util.Set;

@SupportedAnnotationTypes(
    "java.lang.Deprecated"
)
@SupportedSourceVersion(SourceVersion.RELEASE_11)
@AutoService(Processor.class)
public class DeprecatedProcessor extends AbstractProcessor {
    @Override
    public boolean process(Set<? extends TypeElement> annotations, RoundEnvironment roundEnv) {
        for (TypeElement annotation : annotations) {
            for (Element element : roundEnv.getElementsAnnotatedWith(annotation)) {

                processingEnv.getMessager().printMessage(Diagnostic.Kind.ERROR, "Deprecated code should be removed", element);
            }
        }

        return true;
    }
}

File: core/kotlin/bestbefore-processor-kotlin/src/main/kotlin/com/github/gotson/bestbefore/processor/kotlin/DeprecatedProcessor.kt
package com.github.gotson.bestbefore.processor.kotlin

import com.google.auto.service.AutoService
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.processing.SymbolProcessor
import com.google.devtools.ksp.processing.SymbolProcessorEnvironment
import com.google.devtools.ksp.processing.SymbolProcessorProvider
import com.google.devtools.ksp.symbol.KSAnnotated

class DeprecatedKotlinProcessor(private val environment: SymbolProcessorEnvironment) : SymbolProcessor {
  override fun process(resolver: Resolver): List<KSAnnotated> {
    resolver.getSymbolsWithAnnotation("kotlin.Deprecated").forEach {
      environment.logger.error("Deprecated code should be removed", it)
    }
    return emptyList()
  }
}

@AutoService(SymbolProcessorProvider::class)
class DeprecatedProcessorProvider : SymbolProcessorProvider {
  override fun create(environment: SymbolProcessorEnvironment): SymbolProcessor {
    return DeprecatedKotlinProcessor(environment)
  }
}